File: backend/internal/domain/keypair.go
// keypair.go —— Phase C: Owner Ed25519 keypair domain value。私钥永远不
// 入 domain (owner 自己保管 PEM)；本结构只承载 owner 看得到的 metadata +
// backend 验签需要的公钥 PEM。
//
// 撤销 = hard delete (无 status 字段)。对齐 youteacher 极简风格。

package domain

import (
	"errors"
	"time"
)

// ErrKeypairUnauthorized —— 通用 sigv1 验失败 (key 不存在 / sig 不通 /
// ts 超窗口 / header 形态错)。caller 翻 HTTP 401。
var ErrKeypairUnauthorized = errors.New("keypair: unauthorized")

// OwnerKeypair —— DB 一行。PublicKeyPEM 用 PKCS8 Ed25519 编码。
type OwnerKeypair struct {
	LastUsedAt   *time.Time
	CreatedAt    time.Time
	ID           string
	OwnerID      string
	KeyID        string
	PublicKeyPEM string
	Label        string
}

// OwnerKeypairMetadata —— admin list 用：去掉 PEM + ownerID，只剩 owner UI
// 关心的字段。永远不返公钥 (owner 已下载 PEM 自己保管，不需要看 server side)。
type OwnerKeypairMetadata struct {
	LastUsedAt *time.Time
	CreatedAt  time.Time
	ID         string
	KeyID      string
	Label      string
}

File: e2e/fixtures/keypair.ts
// keypair.ts —— Phase C: admin REST helpers for owner_keypairs.
//
// POST /api/admin/keypairs → 一次性返 {keyId, privateKeyPem}
// GET  /api/admin/keypairs → list metadata only (无 pem，无 hash)
// DELETE /api/admin/keypairs/:keyId → hard delete (= revoke)

import type { APIRequestContext } from '@playwright/test';

const BACKEND = process.env['BACKEND_URL'] ?? 'http://localhost:8000';

export interface CreatedKeypair {
  key_id: string;
  private_key_pem: string;
  label: string;
  created_at: string;
}

export interface KeypairView {
  key_id: string;
  label: string;
  created_at: string;
  last_used_at: string | null;
}

export async function createKeypair(
  request: APIRequestContext, csrf: string, label: string,
): Promise<CreatedKeypair> {
  const res = await request.post(`${BACKEND}/api/admin/keypairs`, {
    headers: { 'X-Csrftoken': csrf },
    data: { label },
  });
  if (res.status() !== 201) {
    throw new Error(`create keypair: ${res.status()} ${await res.text()}`);
  }
  return await res.json() as CreatedKeypair;
}

export async function listKeypairs(
  request: APIRequestContext, csrf: string,
): Promise<KeypairView[]> {
  const res = await request.get(`${BACKEND}/api/admin/keypairs`, {
    headers: { 'X-Csrftoken': csrf },
  });
  if (res.status() !== 200) throw new Error(`list keypairs: ${res.status()}`);
  return await res.json() as KeypairView[];
}

export async function deleteKeypair(
  request: APIRequestContext, csrf: string, keyID: string,
): Promise<void> {
  const res = await request.delete(`${BACKEND}/api/admin/keypairs/${keyID}`, {
    headers: { 'X-Csrftoken': csrf },
  });
  if (res.status() !== 204) {
    throw new Error(`delete keypair: ${res.status()} ${await res.text()}`);
  }
}
